osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs

```diff
@@ -102,6 +104,26 @@ protected override void ParseStreamInto(LineBufferedReader stream, Beatmap beatm
             }
         }
 
+        /// <summary>
+        /// Clamp Difficulty settings to be within the normal range.
+        /// </summary>
+        private void applyDifficultyRestrictions(BeatmapDifficulty difficulty)
+        {
+            difficulty.DrainRate = Math.Clamp(difficulty.DrainRate, 0, 10);
+            //If the mode is not Mania, clamp circle size to [0,10]
+            if (!beatmap.BeatmapInfo.Ruleset.OnlineID.Equals(3))
+                difficulty.CircleSize = Math.Clamp(difficulty.CircleSize, 0, 10);
+            //If it is Mania, it must be within [1,20] - dual stages with 10 keys each.
+            //The lower bound should be 4, but there are ranked maps that are lower than this.
+            else
+                difficulty.CircleSize = Math.Clamp(difficulty.CircleSize, 1, 20);
```

Comment:
Technically only [1,18] is supported. Was 20 an intentional choice?

Reply:
May've misread the code in that case - we have `ManiaRuleset.MAX_STAGE_KEYS = 10`, so I assumed that the limit was 20... But ah, I assume allowing when single to have 10 keys, but only allowing 9 when dual would make sense.
I suppose out of scope, but should `for (int i = 2; i <= MAX_STAGE_KEYS * 2; i += 2)` be changed to `for (int i = 2; i <= (MAX_STAGE_KEYS-1) * 2; i += 2)` in that case? 

https://github.com/ppy/osu/blob/3a608aa8153541f4f610d7bea018ce67586f9907/osu.Game.Rulesets.Mania/ManiaRuleset.cs#L327-L336

Reply:
Stable does this:

```csharp
                                        case "CircleSize":
                                            if (BeatmapVersion >= 13)
                                            {
                                                if (PlayMode == PlayModes.OsuMania)
                                                    DifficultyCircleSize = Math.Min(18.0f, Math.Max(1.0f, Single.Parse(right, GameBase.nfi)));
                                                else
                                                    DifficultyCircleSize = Math.Min(10.0f, Math.Max(0.0f, Single.Parse(right, GameBase.nfi)));
                                            }
                                            else
                                                DifficultyCircleSize = Math.Min((byte)10, Math.Max((byte)0, byte.Parse(right)));
```

https://github.com/peppy/osu-stable-reference/blob/0e425c0d525ef21353c8293c235cc0621d28338b/osu!/GameplayElements/Beatmaps/Beatmap.cs#L577-L586

The 18 thing may be coming from [ranking criteria](https://osu.ppy.sh/wiki/en/Ranking_criteria/osu%21mania):

>  Beatmaps can only use 4–10, 12, 14, 16, or 18 keys.

I don't see much harm in supporting 20, albeit there is an argument we should possibly be disallowing odd keycounts above 10 if we're to match rc. Not sure.

Reply:
I believe osu!stable supports odd key counts but not >18K.

Without getting into the weeds about our stance on backwards compatibility and all that, I think we take the easiest path forward and emulate exactly what stable's doing.

Reply:
> Hmm, I receive 404 for the github link...

Yes that's a github link that is supposed to 404 for anyone that's not in the core team.

Reply:
Ahhh, okay. I thought at first it took me to the actual page, so was a bit confused.